### PR TITLE
Simple Game Server: Add \n to Counters and Lists Response

### PR DESCRIPTION
**Pull request by Kalaiselvi84**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, press enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug

/kind cleanup
> /kind documentation
> /kind feature
> /kind hotfix
> /kind release

**What this PR does / Why we need it**:

**Which issue(s) this PR fixes**:

Closes #

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by Kalaiselvi84** [comment]
I only changed `GET_COUNTER_COUNT`. Could you please confirm if this approach is correct? I'm unable to create a fleet with Counters and List, so I couldn't test it fully, but `TestCounters` passed with this change on my local.

File: examples/simple-game-server/main.go [hunk 1]
```diff
@@ -313,7 +313,14 @@ func handleResponse(txt string, s *sdk.SDK, cancel context.CancelFunc) (response
 			responseError = fmt.Errorf("Invalid GET_COUNTER_COUNT, should have 1 arguments")
 			return
 		}
-		response, responseError = getCounterCount(s, parts[1])
+		count, err := getCounterCount(s, parts[1])
+		if err != nil {
```
Comment (igooch) [inline examples/simple-game-server/main.go:317]: In general I'd recommend keeping formatting changes within the getCounterCount method.

**Review by igooch — COMMENTED** [review]
The build will fail until the example simple-game-server image has been updated (otherwise the tests pull from the old image and don't "see" any changes to the simple-game-server/main.go file.) Although we should wait to update the image until all the tests pass locally.

File: examples/simple-game-server/main.go (hunk 1 above)
Comment (Kalaiselvi84) [reply examples/simple-game-server/main.go]: Updated the format changes in getCounterCount() and other methods.

File: test/e2e/gameserver_test.go [hunk 2]
```diff
@@ -1364,10 +1364,10 @@ func TestCounters(t *testing.T) {
 	}{
 		"GetCounterCount": {
 			msg:  "GET_COUNTER_COUNT games",
-			want: "1",
+			want: "COUNTER: 1\n",
 		},
 		"GetCounterCount Counter Does Not Exist": {
-			msg:  "GET_COUNTER_COUNT fame",
+			msg:  "GET_COUNTER_COUNT game",
```
[the remark this answers is not in the document]
Comment (Kalaiselvi84) [reply test/e2e/gameserver_test.go]: I've reverted the text from `game` back to its original `fame`

**Comment by Kalaiselvi84** [comment]
Tests `TestCounters` and `TestLists` have succeeded. Log details - https://gist.github.com/Kalaiselvi84/72d9950899deb9a00eb087e592ebf515

File: test/e2e/gameserver_test.go [hunk 3]
```diff
@@ -1364,93 +1364,93 @@ func TestCounters(t *testing.T) {
 	}{
 		"GetCounterCount": {
 			msg:  "GET_COUNTER_COUNT games",
-			want: "1",
+			want: "COUNTER: 1\n",
```
Comment (markmandel) [inline test/e2e/gameserver_test.go:1367]: To make this fun - you'll need to wait until the new version of simple-game-server is in production before changing the e2e test.

Which also means this PR will need an increment to the example image version 😃 up to you if you want to do it in separate PR's or not.

But otherwise, this looks great 👍🏻

Comment (Kalaiselvi84) [reply test/e2e/gameserver_test.go:1367]: Could you share your preference? I'll go with that🙂

Comment (Kalaiselvi84) [reply test/e2e/gameserver_test.go:1367]: During the testing of some modified methods, I noticed that the changes aren't reflecting properly. I will draft, test, and update the code accordingly.

**Review by igooch — COMMENTED** [review]
Looks good. The only item that looked off was there appeared to be an empty error message for the `GET_LIST_VALUES` when the passed List does not exist.

```
~/agones/examples$ nc -u 34.168.36.96 7477
INCREMENT_COUNTER games 1
SUCCESS: true
GET_COUNTER_COUNT games
COUNTER: 2
DECREMENT_COUNTER sessions 1
SUCCESS: true
GET_COUNTER_COUNT sessions
COUNTER: 0
SET_COUNTER_COUNT sessions 10000
ERROR: false
SET_COUNTER_COUNT sessions 10
SUCCESS: true
GET_COUNTER_COUNT sessions
COUNTER: 10
GET_COUNTER_CAPACITY sessions
CAPACITY: 1000
SET_COUNTER_CAPACITY games 99
SUCCESS: true
GET_COUNTER_CAPACITY games
CAPACITY: 99
GET_LIST_CAPACITY rooms
CAPACITY: 333
SET_LIST_CAPACITY rooms 100
SUCCESS: true
GET_LIST_CAPACITY rooms
CAPACITY: 100
LIST_CONTAINS room1
ERROR: Invalid LIST_CONTAINS, should have 2 arguments
LIST_CONTAINS rooms room1
SUCCESS: true
LIST_CONTAINS rooms room_1
false
ACK: 
GET_LIST_LENGTH rooms
LENGTH: 3
GET_LIST_VALUES rooms
VALUES: room1,room2,room3
APPEND_LIST_VALUE rooms room4
SUCCESS: true
GET_LIST_VALUES abc
ERROR: 

GET_LIST_VALUES rooms
VALUES: room1,room2,room3,room4
DELETE_LIST_VALUE room1
ERROR: Invalid DELETE_LIST_VALUE, should have 2 arguments
DELETE_LIST_VALUE rooms room1
SUCCESS: true
```

**Comment by Kalaiselvi84** [comment]
Right, also the response for the LIST_CONTAINS method should be prepended with `ERROR: ` which is missing.

LIST_CONTAINS rooms room_1
Error: false

**Comment by igooch** [comment]
Could you also update some typos in the readme too? They're missing the second "

https://github.com/googleforgames/agones/blob/e52f20a64b8fa1d20bd8ab1080ca923437794c04/examples/simple-game-server/README.md?plain=1#L42-L47

**Comment by markmandel** [comment]
```
Checking cache file existence for: gs://agones-build-cache/cpp-sdk-build-3337595207.tgz
Downloading cache file: gs://agones-build-cache/cpp-sdk-build-3337595207.tgz...
NotFoundException: 404 gs://agones-build-cache/cpp-sdk-build-3337595207.tgz does not exist.
NotFoundException: 404 gs://agones-build-cache/cpp-sdk-build-3337595207.tgz does not exist.
NotFoundException: 404 gs://agones-build-cache/cpp-sdk-build-3337595207.tgz does not exist.
NotFoundException: 404 gs://agones-build-cache/cpp-sdk-build-3337595207.tgz does not exist.
CommandException: Some components of ./cpp-sdk-build-3337595207.tgz were not downloaded successfully. Please retry this download.
Restoring cache from file ./cpp-sdk-build-3337595207.tgz...
```

That's a weird flake with the caching system.

File: examples/simple-game-server/main.go [hunk 4]
```diff
@@ -787,8 +795,12 @@ func listContains(s *sdk.SDK, listName string, value string) (string, error) {
 	ok, err := s.Alpha().ListContains(listName, value)
 	if err != nil {
 		log.Printf("Error getting List %s contains value %s: %s", listName, value, err)
+		return "ERROR: " + strconv.FormatBool(ok) + "\n", err
 	}
-	return strconv.FormatBool(ok), err
+	if ok {
+		return "SUCCESS: " + strconv.FormatBool(ok) + "\n", nil
+	}
+	return "ERROR: " + strconv.FormatBool(ok) + "\n", err
```
Comment (markmandel) [inline examples/simple-game-server/main.go:803]:
```suggestion
		return "SUCCESS: " + strconv.FormatBool(ok) + "\n", nil
```

I think this would be better - since it's a true/false as to whether if it's found. It's not really an ERROR.

**Review by markmandel — COMMENTED** [review]
If you want to test locally without having to push up to an Agones cluster, you can use the [local SDK Server](https://agones.dev/site/docs/guides/client-sdks/local/#running-the-sdk-server)!

1. `make build-agones-sdk-binary-linux-amd64` (assuming you are on Linux - only need to do this once to ensure you have a local build of the latest local sdk server)
2. Go to `agones/cmd/sdk-server/bin`
3. Run `./sdk-server.linux.amd64 --local --feature-gates=CountsAndLists=true`
4. In another terminal, go to `agones/examples/simple-game-server`
5. `make  build-linux-image-amd64` (since we only need the one image)
6. `docker run --rm --network=host simple-game-server:0.24-linux-amd64` (run on on local host network, so it can connect to the local sdk serer process)
4. In another terminal: `nc -u localhost 7654` and start typing!

I saw one small change to LIST_CONTAINS, but otherwise, this looks good to me.

Some of my testing:

```shell
❯ nc -u localhost 7654
GET_COUNTER_COUNT rooms
COUNTER: 2
GET_LIST_VALUES players
VALUES: test0,test1,test2
LIST_CONTAINS players test1
SUCCESS: true
LIST_CONTAINS players test2
SUCCESS: true
LIST_CONTAINS players NOPE
ERROR: false
```

File: examples/simple-game-server/main.go (hunk 4 above)
Comment (Kalaiselvi84) [reply examples/simple-game-server/main.go]: Done!

**Comment by igooch** [comment]
You'll need to bump the simple-game-server example image and push the updated example image up to the Agones artifact registry first, so that the tests can use the correct image.

File: examples/simple-game-server/main.go (hunk 4 above)
Comment (Kalaiselvi84) [reply examples/simple-game-server/main.go]: `ListContains false` test fails with this change.

Shall we modify it as follows?
```
	if ok {
		return "SUCCESS: " + strconv.FormatBool(ok) + "\n", nil
	} else {
		log.Printf("Error getting List %s contains value %s: %s", listName, value, err)
		return "ERROR: " + strconv.FormatBool(ok) + "\n", err
	}
```

Comment (Kalaiselvi84) [reply examples/simple-game-server/main.go]: Latest commit satisfies the edge cases. Please review it and share your thoughts.

File: test/e2e/gameserver_test.go (hunk 3 above)
Comment (markmandel) [reply test/e2e/gameserver_test.go:1367]: > Could you share your preference? I'll go with that🙂

No preference 😃

File: examples/simple-game-server/main.go [hunk 5]
```diff
@@ -787,8 +795,12 @@ func listContains(s *sdk.SDK, listName string, value string) (string, error) {
 	ok, err := s.Alpha().ListContains(listName, value)
 	if err != nil {
 		log.Printf("Error getting List %s contains value %s: %s", listName, value, err)
+		return strconv.FormatBool(ok), err
+	}	
+	if ok {
+		return "SUCCESS: " + strconv.FormatBool(ok) + "\n", nil
 	}
-	return strconv.FormatBool(ok), err
+	return "ERROR: " + strconv.FormatBool(ok) + "\n", err
```
Comment (markmandel) [inline examples/simple-game-server/main.go:803]: I realised what I think this should be:

```go
return "FOUND: " + strconv.FormatBool(ok) + "\n", nil
```

Which makes way more sense I think?

Also will need to adjust the unit test to match this change.

Comment (Kalaiselvi84) [reply examples/simple-game-server/main.go]: Updated with FOUND in both the `main.go` and `gameserver_test.go` files.

File: examples/simple-game-server/main.go [hunk 6]
```diff
@@ -807,8 +820,12 @@ func getListValues(s *sdk.SDK, listName string) (string, error) {
 	values, err := s.Alpha().GetListValues(listName)
 	if err != nil {
 		log.Printf("Error getting List %s values: %s", listName, err)
+		return strings.Join(values, ","), err
 	}
-	return strings.Join(values, ",") + "\n", err
+	if len(values) > 0 {
+        return "VALUES: " + strings.Join(values, ",") + "\n", nil
+    }
+	return "Empty List" + strings.Join(values, ",") + "\n", err
```
Comment (Kalaiselvi84) [inline examples/simple-game-server/main.go:828]: Could you review the getListValues method? I'm unsure if anything is missing. In negative test cases, it only returns 'ERROR:' without further details. Is this behavior expected?

Comment (igooch) [reply examples/simple-game-server/main.go:828]: It's normal for the examples/simply-game-server/main.go code. This code only sends back the return value, not any return error. And in the case of `GetListValues` the return value is nil. We probably should pipe through the error message, but that's not in the code as it stands. 

https://github.com/googleforgames/agones/blob/12159252599bf2407cc3f666796a39a9ce4b8eb3/examples/simple-game-server/main.go#L496-L511 is where it only returns the response.

Comment (markmandel) [inline examples/simple-game-server/main.go:828]:
```suggestion
	return "VALUES: <none>\n", err
```
Or something like that? To be consistent with the other responses of having a "DESCRIPTOR:RESULT" format.

**Review by igooch — COMMENTED** [review]
It looks like the updated image reference is missing in a couple places like `test/e2e/framework.go`.

**Comment by Kalaiselvi84** [comment]
> It looks like the updated image reference is missing in a couple places like `test/e2e/framework.go`.

Will update the new image in the `test/e2e` directory once the image pushed to prod(doing it now).

**Comment by Kalaiselvi84** [comment]
A new image has been pushed for the simple-game-server in prod, and the necessary files have been updated with this new image.

**Comment by Kalaiselvi84** [comment]
I've pushed a new image to the simple-game-server in production. However, I'm not really sure about the reason for the build failure. Am I missing anything that needs updating? If we can't fix this build issue, should we delete the new tag from the simple-game-server since tomorrow is the release day?

**Comment by Kalaiselvi84** [comment]
Log for passed tests - https://gist.github.com/Kalaiselvi84/4196d647e4c069bbe3b1237e8033ca66

File: examples/simple-game-server/main.go [hunk 7]
```diff
@@ -787,8 +795,12 @@ func listContains(s *sdk.SDK, listName string, value string) (string, error) {
 	ok, err := s.Alpha().ListContains(listName, value)
 	if err != nil {
 		log.Printf("Error getting List %s contains value %s: %s", listName, value, err)
+		return strconv.FormatBool(ok), err
+	}	
+	if ok {
+		return "FOUND: " + strconv.FormatBool(ok) + "\n", nil
```
Comment (markmandel) [inline examples/simple-game-server/main.go:801]: I think this review comment got lost somewhere - but this result should be the following.

```go
func listContains(s *sdk.SDK, listName string, value string) (string, error) {
	log.Printf("Getting List %s contains value %s", listName, value)
	ok, err := s.Alpha().ListContains(listName, value)
	if err != nil {
		log.Printf("Error getting List %s contains value %s: %s", listName, value, err)
		return strconv.FormatBool(ok), err
	}	
  return "FOUND: " + strconv.FormatBool(ok) + "\n", nil
}
```

Not finding a value shouldn't be an error - it's just a search function.

File: test/e2e/gameserver_test.go [hunk 8]
```diff
@@ -1525,69 +1525,69 @@ func TestLists(t *testing.T) {
...
 		},
 		"ListContains": {
 			msg:  "LIST_CONTAINS games game2",
-			want: "true",
+			want: "FOUND: true\n",
 		},
 		"ListContains false": {
 			msg:  "LIST_CONTAINS games game0",
-			want: "false",
+			want: "ERROR: false\n",
```
Comment (markmandel) [inline test/e2e/gameserver_test.go:1554]: This test will need to change with the update to the response.

**Review by markmandel — COMMENTED** [review]
Congrats on getting this to pass!

I think one of my review comments kept getting lost, so I tried to make it as clear as possible. I think that's about it on review though. Lemme know if that's not clear. Shouldn't need to delete 0.26, since the current release isn't using it.

Since K8s clusters will cache existing image tags for a while, so once the review change has made, we will need to increment the version number and push again, as we can't force remove it from all clusters.

File: test/e2e/gameserver_test.go (hunk 8 above)
Comment (Kalaiselvi84) [reply test/e2e/gameserver_test.go]: Sorry, I thought the response for `ListContains false` should be `ERROR: false`. I have now updated it as per your suggestion. The response for the `ListContains` is `FOUND: true` and for `ListContains false` it is `FOUND: false`.

Comment (markmandel) [reply test/e2e/gameserver_test.go]: 👍🏻 no worries - had a feeling my suggestion wasn't 100% clear. I feel like this will break the e2e test, since a new image.

This all LGTM! We just need it to pass! Probably best bet is to increment the tag again everywhere and push a new image.

Comment (Kalaiselvi84) [reply test/e2e/gameserver_test.go]: Will increment the tag now

Comment (Kalaiselvi84) [reply test/e2e/gameserver_test.go]: I've pushed the new image to the agones-images project and updated the necessary files with the new tag.

Comment (markmandel) [reply test/e2e/gameserver_test.go]: 🤞🏻

Comment (Kalaiselvi84) [reply test/e2e/gameserver_test.go]: Success! The build has finished without any issues.

**Review by markmandel — APPROVED** [review]
LGTM! @igooch giving you a last right of refusal, but if you are happy, I'm happy to merge just in time before release 👍🏻
